docs/Changelog.md

```diff
@@ -18587,6 +18587,95 @@ This version of the operator has been available since version 13 of the default
 </dl>
 
 ## Version 14 of the default ONNX operator set
+### <a name="BatchNormalization-14"></a>**BatchNormalization-14**</a>
+
+  Carries out batch normalization as described in the paper
+  https://arxiv.org/abs/1502.03167. Depending on the mode it is being run,
+  There are five required inputs 'X', 'scale', 'B', 'input_mean' and
+  'input_var', in addition to one optional input 'training_mode'.
+  Note that 'input_mean' and 'input_var' are expected to be the estimated
+  statistics in inference mode (training_mode=False, default),
+  and the running statistics in training mode (training_mode=True).
+  There are multiple cases for the number of outputs, which we list below:
+  
+  Output case #1: Y, running_mean, running_var, current_mean, current_var (training_mode=True)
+  Output case #2: Y (training_mode=False)
+  
+  When training_mode=False, extra outputs are undefined and the user should not depend on those.
```

Comment:
A note about this as it relates to ORT implementation: if the extra outputs are specified and ignored by the ORT kernel, it is possible that even the memory for these output tensors are not allocated. This can cause subsequent errors with unallocated memory usage (depending on how the memory-planner reuses these buffers). So, it may be preferable to flag the use of extra outputs as error when training_mode=False. It would be helpful to get some input from ORT side on this. (@askhade any thoughts?)

Reply:
Here is the ORT implementation: https://github.com/microsoft/onnxruntime/blob/8892ee4b6d343109699ab292e66c2c7a5e41925a/onnxruntime/core/providers/cpu/nn/batch_norm.h#L78

I don't think the backend needs to access memory location of such outputs in inference.

Reply:
cc @pranavsharma

Reply:
@neginraoof While the batch_norm kernel doesn't access the optional tensors in inference mode, I believe the concern was over the allocation planner. I'm in favor of making it invalid to use the extra outputs when in inference mode, since it reduces edge-cases and is generally confusing anyway (why introduce the notion of "undefined behavior" when you can just prevent i t entirely)?

Reply:
Btw the discussion in https://github.com/onnx/onnx/issues/1042 might be relevant here. I think the consensurs (at least for n=2, @gramalingam and I) was that there is not really any use-case for the optional outputs during inference. While this means that we could technically change the spec so that training vs. inference mode is determined purely based on presence of these optional outputs, I don't think changing the behavior of an op based on # of outputs is as clean as just having an attribute for it.

Reply:
One challenge is that the "mode" is an input (as opposed to an attribute). I guess this was chosen to allow a training-engine to use the same graph/node, but change the mode dynamically? If we want this ability, then we will have to allow the multiple outputs to be specified. This also means that the issue mentioned above for allocation-planner buffer-reuse is a problem and ORT will need to solve it someway. A trivial fix would be for the kernel to ensure that the output is allocated by calling "Output" even if it's value is not initialized, but this would mean some unnecessary memory allocation in some situations.

Reply:
Oh I missed that `mode` was an input instead of attribute. IIRC in the older (reverted) PR it was an attribute instead. @neginraoof can you clarify the motivation for it being an input?

> allow a training-engine to use the same graph/node, but change the mode dynamically
I don't think this is an issue because ORT already has a pre-training graph transform pass that can modify attributes or add outputs. Being able to change mode by setting a graph input doesn't seem like it provides any concrete advantage over just changing the attribute.

Reply:
Yes, an attribute would simplify things. It would be great to clarify the motivation to have it as an input, thanks!

Reply:
This is the revert PR: https://github.com/onnx/onnx/pull/2750/files
This one shows that training_mode has been an input for the op before, and I kept it based on the old implementations.
But from your feedback, looks like we can do 2 things:
1- Make training_mode an attribute.
2- Indicating extra outputs are invalid when training_mode=False

Reply:
Ah I see. If it were solely up to me, I think that having it as an attribute is cleaner – and for consistency it's dropout that should be updated to move the "training_mode" to an attribute (there have been various other ops in the past that have had inputs moved as attributes, so this is not really anything unprecedented – and seems to indicate to me that having things as attributes is what schemas generally evolve towards anyway). But I'm just a user so I'll defer to the others on whatever they think is better.